content/blog/using-the-gdb4hpc-debugger-at-scale-with-a-cuda-mpi-hpc-application.md

```diff
@@ -0,0 +1,545 @@
+This is free software: you are free to change and redistribute it.
+There is NO WARRANTY, to the extent permitted by law.
+```
+
+#### Launch the Application in gdb4hpc
+
+Start gdb4hpc. You will be dropped into a command line interface. gdb4hpc is operated
+by issuing debugging commands on the command line. The commands should be familiar
+to anyone who has used gdb before.
+
+```shell
+$ gdb4hpc
+gdb4hpc 4.16.3. - Cray Interactive Parallel Debugger
+With Cray Comparative Debugging Technology.
+Copyright 2007-2024 Hewlett Packard Enterprise Development LP.
+Copyright 1996-2016 University of Queensland. All Rights Reserved.
+
+Type "help" for a list of commands.
+Type "help <cmd>" for detailed help about a command.
+dbg all>
+```
+
+Use the `launch` command to start the application in the debugger.
+
+```shell
+dbg all> launch $simpleMPI{8} --launcher-args="-N2 --ntasks-per-node=4 --exclusive -p griz256" --gpu ./simpleMPI
+Starting application, please wait...
+Launched application...
+0/8 ranks connected... (timeout in 300 seconds)
+..
+8/8 ranks connected.
+Created network...
+Connected to application...
+Launch complete.
+simpleMPI{0..7}: Initial breakpoint, main at simpleMPI.cpp:63
+```
+
+`$simpleMPI{8}` tells gdb4hpc to launch a job with 8 ranks and give it a
+handle called `simpleMPI`.
+
+`--launcher-args` tells gdb4hpc which WLM arguments to use to launch the job
+correctly. Use the same arguments you used in the **Test The
+Application** section.
+
+The `--gpu` flag tells gdb4hpc to use the appropriate GPU-aware debugger. Here, that's cuda-gdb.
+
+The final argument, `./simpleMPI`, specifies the binary to launch.
+
+After the launch completes, gdb4hpc stops at `main` and gives you control.
+
+#### Stop at a Breakpoint in a GPU kernel
+
+The sample application contains one CUDA Kernel called `simpleMPIKernel`
+defined in simpleMPI.cu. The source code for the kernel looks like this:
+
+```cpp
+52  // Device code
+53  // Very simple GPU Kernel that computes square roots of input numbers
+54  __global__ void simpleMPIKernel(float *input, float *output) {
+55    int tid = blockIdx.x * blockDim.x + threadIdx.x;
+56    output[tid] = sqrt(input[tid]);
+57  }
+```
+
+Set a breakpoint at the first line of the kernel with the `break` command, just
+like you would in gdb:
+
+```shell
+dbg all> break simpleMPI.cu:55
+simpleMPI{0..7}: Breakpoint 1: file simpleMPI.cu, line 55.
+```
+
+> You could have also set the breakpoint via the function name by typing `break
+> simpleMPIKernel`.
+
+Now use the `continue` command to run the application until it reaches the
+breakpoint in the kernel.
+
+```shell
+dbg all> continue
+dbg all> <$simpleMPI>: Running on 8 nodes
+simpleMPI{0..7}: Breakpoint 1,  at simpleMPI.cu:55
+```
+
+The application has stopped in a GPU kernel. From here you can use the usual
+debugger commands like `print`, `step`, `next`, etc. You can use `print` to
+inspect GPU memory as you would any other memory. Here we print the first 8 items
```

Comment:
Modify this sentence that currently reads "Here we print the first 8 items..." so you are not using the Royal We. Try this instead: "Print the first 8 items of the input array."